File: Round-7/多项式族/jury.java
import java.util.Scanner;

public class jury {
	static class Point {
		int x, y;

		Point(int x, int y) {
			this.x = x;
			this.y = y;
		}
	}

	static double interpolate(int x, Point[] f) {
		for (int i = 0; i < f.length; i++) {
			if (f[i].x == x) {
				return f[i].y;
			}
		}
		double res = 0;
		for (int i = 0; i < f.length; i++) {
			double num = 1, den = 1;
			for (int j = 0; j < f.length; j++) {
				if (i != j) {
					num *= x - f[j].x;
					den *= f[i].x - f[j].x;
				}
			}
			res += f[i].y * num / den;
		}
		return res;
	}

	static boolean doubleEquals(double a, double b) {
		return Math.abs(a - b) / Math.max(1, Math.max(Math.abs(a), Math.abs(b))) < 1e-6;
	}

	public static void main(String[] args) {
		Scanner sc = new Scanner(System.in);
		int n = sc.nextInt();
		Point[] f = new Point[n + 1], g = new Point[n + 1];
		for (int i = 0; i <= n; i++) {
			f[i] = new Point(sc.nextInt(), sc.nextInt());
		}
		for (int i = 0; i <= n; i++) {
			g[i] = new Point(sc.nextInt(), sc.nextInt());
		}
		double c = f[0].y - interpolate(f[0].x, g);
		for (int i = 1; i <= n; i++) {
			if (!doubleEquals(interpolate(f[i].x, g) + c, f[i].y)) {
				System.out.println("NO");
				return;
			}
		}
		System.out.println("YES");
		System.out.printf("%.10f\n", c);
	}
}

File: Round-7/多项式族/jury.c
#include <stdio.h>
#include <math.h>
#include <stdbool.h>

struct Point {
	int x, y;
};

static double interpolate(int x, int n, struct Point points[]) {
	for (int i = 0; i < n; i++) {
		if (points[i].x == x) {
			return points[i].y;
		}
	}
	double y = 0;
	for (int i = 0; i < n; i++) {
		double t = points[i].y;
		for (int j = 0; j < n; j++) {
			if (i != j) {
				t *= ((double)x - points[j].x) / (points[i].x - points[j].x);
			}
		}
		y += t;
	}
	return y;
}

static bool dbl_eq(double x, double y) {
	return fabs(x - y) / fmax(fmax(fabs(x), fabs(y)), 1.0) <= 1e-6;
}

int main(void) {
	int n;
	scanf("%d", &n);
	struct Point f[n + 1], g[n + 1];
	for (int i = 0; i <= n; i++) {
		scanf("%d%d", &f[i].x, &f[i].y);
	}
	for (int i = 0; i <= n; i++) {
		scanf("%d%d", &g[i].x, &g[i].y);
	}
	double c = f[0].y - interpolate(f[0].x, n + 1, g);
	for (int i = 1; i <= n; i++) {
		if (!dbl_eq(f[i].y, interpolate(f[i].x, n + 1, g) + c)) {
			puts("NO");
			return 0;
		}
	}
	puts("YES");
	printf("%.10f\n", c);
	return 0;
}